src/os_net/os_net.c

```diff
@@ -40,92 +45,173 @@ static socklen_t us_l = sizeof(n_us);
 #endif /* WIN32*/
 
 
-/* Bind a specific port */
-int OS_Bindport(char *_port, unsigned int _proto, const char *_ip)
+/* Bind all relevant ports */
```

Comment:
Does this change affect the `local_ip` option? 
http://www.ossec.net/docs/syntax/head_ossec_config.remote.html#element-local_ip

Reply:
No, it does not affect the local_ip option - it continues to work the same way is always did. The OSNetInfo() function in os_net/os_net.c receives the port and IP address as character pointers (text strings) and the connection prototype (TCP or UDP) as an integer. Those values are passed to getaddrinfo() in the same way they were passed in the original code. However, in the absence of the local_ip being specified, this change will allow it to bind to all active IP addresses instead of the first address it finds in the system interface list.

There are several advantages to this approach. First, under the old code, you got the first address getaddrinfo() found. On FreeBSD, if that first address was not your primary interface address (i.e., receiving an IPv6 address when you are configured for an IPv4 network), you had to disable IPv6 to get OSSEC to work (which broke other things). Second, the old code did not support multiple interfaces (like multiple ethernet ports on pfSense routers) - this new code does. Third, the old code did not let you support both IPv6 and IPv4 on the same interface (although, with the old code you could do "mapped" IPv4 over IPv6). Under the new code, this is no longer an issue - both IPv4 and IPv6 are supported at the same time.

Note that I intentionally left the Linux code as it was because it was working without issues (but for just one address). However, it might be beneficial to use AF_UNSPEC in place of AF_INET6 for the Linux code too. I added the ability for Linux to support AF_UNSPEC with a #define NOV4MAP. Allowing all systems running OSSEC to support multiple IP addresses using both IPv4 and IPv6 and/or multiple interfaces is really useful.

I have this code running on seven FreeBSD 11.1 and 11.2 systems and one pfSense system and it is working like a charm. I also have a CentOS Linux system in my office that has been running the modified code without issues. I have seen no performance issues with the use of select() in the new code for handling the incoming messages.

Reply:
Thank you for the detailed reply. Since 3.0 has finally dropped, I've finally gotten a chance to play with this. I haven't seen any problems so far. I'll try to give it a shot with Linux/NOV4MAP.